File: api/match-service/src/app/routes/matrix.py
from flask import Blueprint, jsonify, request
from app.services.matrix_service import MatrixService

matrix_bp = Blueprint('matrix', __name__, url_prefix='/api/v1/matrix')

# Initialize matrix service
matrix_service = MatrixService()

@matrix_bp.route("/users", methods=["GET"])
def get_users_matrix():
    """Get users data in matrix format for matching calculations"""
    try:
        # Get optional user_ids parameter
        user_ids_param = request.args.get('user_ids')
        user_ids = None
        
        if user_ids_param:
            try:
                user_ids = [int(uid.strip()) for uid in user_ids_param.split(',')]
            except ValueError:
                return jsonify({
                    "success": False,
                    "error": "Invalid user_ids format. Use comma-separated integers."
                }), 400
        
        # Get include_metadata parameter (default True)
        include_metadata = request.args.get('include_metadata', 'true').lower() == 'true'
        
        # Convert users to matrix format
        matrix_data = matrix_service.get_users_matrix(user_ids, include_metadata)
        
        return jsonify({
            "success": True,
            "data": matrix_data
        })
        
    except Exception as e:
        return jsonify({
            "success": False,
            "error": f"Failed to generate matrix: {str(e)}"
        }), 500

@matrix_bp.route("/compatible/<int:user_id>", methods=["GET"])
def get_compatible_matrix(user_id):
    """Get matrix of users compatible with specified user"""
    try:
        compatible_users = matrix_service.get_compatible_users_matrix(user_id)
        
        return jsonify({
            "success": True,
            "data": {
                "target_user_id": user_id,
                "compatible_users": compatible_users,
                "count": len(compatible_users)
            }
        })
        
    except Exception as e:
        return jsonify({
            "success": False,
            "error": f"Failed to get compatible users: {str(e)}"
        }), 500

@matrix_bp.route("/export", methods=["POST"])
def export_matrix():
    """Export users matrix to file"""
    try:
        data = request.get_json() or {}
        
        # Get parameters from request
        user_ids = data.get('user_ids')
        include_metadata = data.get('include_metadata', True)
        filename = data.get('filename', 'users_matrix.json')
        
        # Export matrix
        result = matrix_service.export_matrix_to_file(user_ids, include_metadata, filename)
        
        return jsonify({
            "success": True,
            "data": result
        })
        
    except Exception as e:
        return jsonify({
            "success": False,
            "error": f"Failed to export matrix: {str(e)}"
        }), 500

File: api/match-service/src/handlers/matrix.go
package handlers

import (
	"net/http"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"

	"match-service/src/services"
	"match-service/src/utils"
)

func GetMatrixHandler(c *gin.Context) {
	var userIDs []int
	
	// Parse user_ids parameter if provided
	if userIDsParam := c.Query("user_ids"); userIDsParam != "" {
		idStrings := strings.Split(userIDsParam, ",")
		for _, idStr := range idStrings {
			if id, err := strconv.Atoi(strings.TrimSpace(idStr)); err == nil {
				userIDs = append(userIDs, id)
			} else {
				utils.RespondError(c, "Invalid user_ids format. Use comma-separated integers.", http.StatusBadRequest)
				return
			}
		}
	}

	// Parse include_metadata parameter (default true)
	includeMetadata := c.DefaultQuery("include_metadata", "true") == "true"

	matrixService := services.NewMatrixService()
	matrixData, err := matrixService.GetUsersMatrix(userIDs, includeMetadata)
	if err != nil {
		utils.RespondError(c, "Failed to generate matrix: "+err.Error(), http.StatusInternalServerError)
		return
	}

	utils.RespondSuccess(c, http.StatusOK, matrixData)
}

func GetCompatibleMatrixHandler(c *gin.Context) {
	userIDStr := c.Param("user_id")
	userID, err := strconv.Atoi(userIDStr)
	if err != nil {
		utils.RespondError(c, "Invalid user ID", http.StatusBadRequest)
		return
	}

	matrixService := services.NewMatrixService()
	compatibleUsers, err := matrixService.GetCompatibleUsersMatrix(userID)
	if err != nil {
		utils.RespondError(c, "Failed to get compatible users: "+err.Error(), http.StatusInternalServerError)
		return
	}

	utils.RespondSuccess(c, http.StatusOK, gin.H{
		"target_user_id":   userID,
		"compatible_users": compatibleUsers,
		"count":           len(compatibleUsers),
	})
}

func GenerateMatrixHandler(c *gin.Context) {
	var request struct {
		UserIDs         []int  `json:"user_ids"`
		IncludeMetadata bool   `json:"include_metadata"`
		Filename        string `json:"filename"`
	}

	if err := c.ShouldBindJSON(&request); err != nil {
		utils.RespondError(c, "Invalid request: "+err.Error(), http.StatusBadRequest)
		return
	}

	if request.Filename == "" {
		request.Filename = "users_matrix.json"
	}

	matrixService := services.NewMatrixService()
	result, err := matrixService.ExportMatrixToFile(request.UserIDs, request.IncludeMetadata, request.Filename)
	if err != nil {
		utils.RespondError(c, "Failed to export matrix: "+err.Error(), http.StatusInternalServerError)
		return
	}

	utils.RespondSuccess(c, http.StatusOK, result)
}